### ISSUE TITLE
MS Sql server as external database

### ISSUE THREAD

**Issue by ghost** [gitbucket/gitbucket#1812]
Does Gitbucket support MS SQL server database to store all the data?

**Comment by aadrian** [comment]
@jimAC2016 No.
As you can see here: https://github.com/gitbucket/gitbucket/wiki/External-database-configuration only those 3 DBs are supported.

**Comment by jimAC2016** [comment]
Will it be added in one of the next features?

**Comment by aadrian** [comment]
> Will it be added in one of the next features?

@jimAC2016 what would be the added value? 
I have quite a few GitBucket instances all running with the embedded H2 database since years without problems.

MySql and Postgres support was added because some users had problems with H2 (I suppose in less reliable hardware setups with embedded DBs) and they are also open source.

**Comment by uli-heller** [comment]
@jimAC2016 Do you want to see an official support for MS SQL Server? Or is it good enough for you to get it up and running? If you just want to get it up and running, then maybe this is enough:

* unpack the war file
* add the jdbc driver of ms sql server to .../WEB-INF/lib
* create a new war file
* configure the database url to use the ms sql server
* start the war file

Maybe you're lucky and gitbucket runs fine...

**Comment by rsingh25** [comment]
MS SQL support cannot be added just by packaging the jdbc driver and changing db url. 

The supported database types are hardcoded in [DatabaseConfig.scala](https://github.com/gitbucket/gitbucket/blob/224c355d44f58aa5fab67df109d8d5f9268f5b15/src/main/scala/gitbucket/core/util/DatabaseConfig.scala#L74). It supports only h2, mysql and postgres.

**Comment by aperiooculus** [comment]
I would like to official support for MS SQL Server databases.

The UI recommends storing the data in an external database if using GitBucket for important purposes. F

![image](https://user-images.githubusercontent.com/7217257/37465972-fbdcde40-2832-11e8-8eef-edf7a5495559.png)

My organization uses MS SQL Server for our development and production environments. Adding support for MS SQL Server would allow us to maintain that alignment.

**Comment by ardanzap** [comment]
Hi, 

I'm trying to configure gitbucket in one of my servers but my instance of MySQL is in another one. Do I have to install it in the same server or can I configure it remotely?

Thanks in advance!

Paula.

PS. I agree with the others, it will be nice to have a MS SQL Server database support too.

**Comment by aadrian** [comment]
@ardanzap 

This is an **MS SQL Server** Topic, not a MySQL one :) .

Please use for support: the chat room, as stated on the main page: https://github.com/gitbucket/gitbucket#support 

> I'm trying to configure gitbucket in one of my servers but my instance of MySQL is in another one. Do I have to install it in the same server or can I configure it remotely?

As it is visible from the docs:
https://github.com/gitbucket/gitbucket/wiki/External-database-configuration#mysql-57-or-higher
with any external database, the connection has a standard (JDBC) **URL**, so as consequence, the DB can be anywhere as long it's reachable under that URL (host and port).

**Comment by aperiooculus** [comment]
Can we mark this as a requested feature or a possible improvement?

**Comment by yaroot** [comment]
Unpacking the war is unnecessary, you can use `-cp` for that. But it won't work. However if you have some hours to kill and want to contribute, you can:

- add drivers and modify `gitbucket.core.util.DatabaseType` to add sqlserver support
- run gitbucket and fix all sql errors if there are any
- make a pull request

It's pretty viable, mssql is free for non-production use and there is a docker image, slick support no longer requires lightbend subscription.

However there's maintenance overhead, but the testing can be automated via CI.

@takezoe please chip in

**Comment by aadrian** [comment]
-1 for for supporting MS SQL Server as an aditional DB .

GitBucket is targeted towards small projects and teams that can't afford (or are not allowed) to use something like GitHub Enterprise (or the concurrent products) - hence the efforts to be API compatible.

For the GitBucket users it's **irrelevant** what persistence is being used.

H2 (in embedded mode) is more than enough most of the time and requires no maintenance.
Since some users seems to have had some problems in the past (most of the time it was IMHO due to missunderstanding how H2 and embedded DBs were working), MySQL and Postregs were added as extra persistence - for the case one would have "problems".

Adding MS SQL Server supports adds **ZERO** value to this project, and only consumes developement and test resources (considering how many open issues exist that would **really** benifit many or most users).

Just my 2 cents.

**Comment by McFoggy** [comment]
I would tend to agree with @aadrian ; even if I would not say it is irrelevant to support external DBs I would really prefer to see efforts in platform stability (including non regression testing)/plugin registry/automatic backup that affects a lot more of project users than the support of MariaDB/MsSqlServer/MySQL/Oracle/DB2 or whatever other DB you can imagine.

Ideally, DB support should be externalized to plugins and gitbucket should come with a default one (H2 probably). This way it could reduce the maintenance/testing effort of the core gitbucket platform.

So -1 for me for a core integration of MsSQL (even with a PR) but +1 to support DBs by plugins.

**Comment by yaroot** [comment]
It would be pretty hard for a plugin to do it, if someone can afford sqlserver he surely could just use one of the db that gitbucket supports.

**Comment by SIkebe** [comment]
@aadrian 
>H2 (in embedded mode) is more than enough most of the time and requires no maintenance ... MySQL and Postregs were added as extra persistence - for the case one would have "problems".

Just to clarify, GitBucket recommends not to use embedded H2. It shows below warning if you use it.
>Your GitBucket is running on embedded H2 database. Recommend to configure to use external database if you would like to use GitBucket for important purpose.

Though MySQL and PostgreSQL support is enough in most cases, adding SQL Server support is not **ZERO** value. Especially for users who already use it. 
Just too costly in terms of maintenance and testing.

**Comment by McFoggy** [comment]
> It would be pretty hard for a plugin to do it

for sure it is not free to do ; it requires a refactoring of the core to allow it.

**Comment by aadrian** [comment]
@SIkebe 

> adding SQL Server support is not ZERO value. Especially for users who already use it.

It is, since users of GitBucket should not interact with the DB that GitBucket is using - it's only there for GitBucket itself.
It is just random that in current versions it's using a relational DB - very well it could have used a NoSQL database or some XMLs or other files or any otherr kind of serialization for it's persistence. 

> Just to clarify, GitBucket recommends not to use embedded H2.

The wording is rather new and IMHO quite unfortunate. Some admins didn't understood how embedded DBs work vs standalone ones, so to avoid the constant support questions of users that don't RTFM it was chosen to "recommend" an external DB.

**Comment by SIkebe** [comment]
@aadrian 
>It is, since users of GitBucket should not interact with the DB that GitBucket is using - it's only there for GitBucket itself.

It is true for you but not always for others. Every decision has its own pros and cons. In this case, I think the cons win.
GitBucket is not a managed cloud service like GitHub. It requires someone who care about data persistence for production usage. 

>It is just random that in current versions it's using a relational DB - very well it could have used a NoSQL database or some XMLs or other files or any otherr kind of serialization for it's persistence.

I agree with you in this regard. Really appreciate if you'd make GitBucket free from the way of persisting data in a realistic manner.

**Comment by yaroot** [comment]
Technically adding drivers & configs would be the easiest part of supporting another database, you'll also need to be fixing all sql in gitbucket and plugins without breaking existing installations. I don't think it would be hard, but these effort could be better spend in improving the existing code base.

Should this be closed as _won't fix_?

**Comment by takezoe** [comment]
I think MS SQL server support is meaningful for some users and adding it is not so difficult as @yaroot mentioned, but keeping maintain it will cause the additional cost continuously and I don't want it.

I agree with closing this issue with "wontfix", but I will not stop that someone forks and adds MS SQL server support to GitBucket. I rather hope it because it's open source!

**Comment by McFoggy** [comment]
> but keeping maintain it will cause the additional cost continuously and I don't want it.

fully understandable

> but I don't stop that someone forks and adds MS SQL server support to GitBucket. I rather hope it because it's open source!

but maintaining a fork with core changes is a pain for the maintainer and can be so costly that few guys will do it. Moreover it is far more difficult to find a goof fork suitable for your need than finding a plugin.

That's why IMHO it would be a good tradeoff to split the effort:
- core team make the DB provider pluggable and create plugins for the core supported DB
- community provide MsSQL server or any other DB support

**Comment by aadrian** [comment]
> core team make the DB provider pluggable and create plugins for the core supported DB

IMO this is a  very good idea :) .

In fact if more functionality in the GitBucket core would be plugg-able, this would allow much more issues from the current issue list to be realized as community plug-ins.

**Comment by joshuacurtiss** [comment]
Allow me to express my viewpoint, and understand there may be others with the same viewpoint, and I preface by expressing my immense gratitude for this project. It is fantastic and invaluable. 

Some institutions like mine have a big fat Microsoft license that includes MSSQL and many other needs. However, they are not so open to spending for other developer needs like source control and project management tools, so I am very happy when I find practical high quality open source options like GitBucket. Furthermore, they're not agreeable to spinning up other database servers (even if they are open source) since MSSQL is their standard.

So I've used H2 since starting our usage of GitBucket for 2 years without incident.  I've ignored the strong worded warning on the admin page about using external database for important purposes and just kept paranoid-level backups.

But combine it with [strong recommendations](https://gitter.im/gitbucket/gitbucket?at=5c178fb5e806954a9a2ff182) in Gitter discussion when [things go wrong](https://gitter.im/gitbucket/gitbucket?at=5c175b93e806954a9a2ea738) and I feel unsafe.

Just sharing, because I find the assumption that anyone using MSSQL will not be interested in GitBucket to be very erroneous from my point of view. This project is very high quality and is a candidate for consideration from companies that utilize MSSQL, and more so if support was added for MSSQL. Thanks for your reading.
